src/service/storage_arango.py

```diff
@@ -0,0 +1,103 @@
+"""
+A storage system for collections based on an Arango backend.
+"""
+
+from aioarango.database import StandardDatabase
+from aioarango.exceptions import CollectionCreateError
+from src.service.models import SavedCollection
+
+
+# service collection names that aren't shared with data loaders.
+_COLL_PREFIX = "kbcoll_coll_"
+_COLL_COUNTERS = _COLL_PREFIX + "counters"
+_COLL_VERSIONS = _COLL_PREFIX + "versions"
+_COLL_ACTIVE = _COLL_PREFIX + "active"
+
+_FLD_KEY = "_key"
+_FLD_COLLECTION = "collection"
+_FLD_COLLECTION_ID = "collection_id"
+_FLD_COUNTER = "counter"
+
+_QUERY_GET_NEXT_VERSION = f"""
+    UPSERT {{{_FLD_KEY}: @{_FLD_COLLECTION_ID}}}
+        INSERT {{{_FLD_KEY}: @{_FLD_COLLECTION_ID}, {_FLD_COUNTER}: 1}}
+        UPDATE {{{_FLD_COUNTER}: OLD.{_FLD_COUNTER} + 1}}
+        IN @@{_FLD_COLLECTION}
+        OPTIONS {{exclusive: true}}
+        RETURN NEW
+"""
+
+# TODO SCHEMA may want a schema checker at some point... YAGNI for now. See sample service
+
+# Assume here that we're never going to make an alternate implementation of this interface.
+# Seems incredibly unlikely
+
+async def create_storage(
+        db: StandardDatabase,
+        create_collections_on_startup: bool = False):
+    """
+    Create the wrapper.
+
+    db - the database where the data is stored. The DB must exist.
+    create_collections_on_startup - on starting the wrapper, create all collections and indexes
+        rather than just checking for their existence. Usually this should be false to
+        allow for system administrators to set up sharding to their liking, but auto
+        creation is useful for quickly standing up a test service.
+    """
+    if create_collections_on_startup:
+        # TODO DB create indexes and collections
+        await _create_collection(db, _COLL_COUNTERS)  # no indexes necessary
+    else:
+        await _check_collection_exists(db, _COLL_COUNTERS)
+        # TODO DB check collections and indexes exist
+    return ArangoStorage(db)
+
+async def _check_collection_exists(db: StandardDatabase, col_name: str):
+    if not await db.has_collection(col_name):
+        raise ValueError(f"Collection {col_name} does not exist")
+
+async def _create_collection(db: StandardDatabase, col_name: str):
+    """
+    Create a collection, ignoring a duplicate name error
+    """
+    try:
+        await db.create_collection(col_name)
+    except CollectionCreateError as e:
+        if not e.error_code == 1207:  # duplicate name error code
```

Comment:
it's ok. the error code is for `name duplicate`. but why not check collection existence before calling create_collection? That will be easier to read than looking for the error code.

Reply:
Because since it's two steps, it'd be a race condition, like how you shouldn't check for a file's existence and then open the file. I'd prefer it if the API allowed a `silent_if_exists` or something option, but there isn't one.

It's not a race condition that is likely to occur, but still, better to keep to best practices.

Reply:
https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use

Reply:
Also just one http connection to arango instead of 2